language-server/src/main/java/org/sonarlint/languageserver/RuleDescriptionHttpServer.java

```diff
@@ -0,0 +1,113 @@
+import org.sonarsource.sonarlint.core.client.api.common.RuleDetails;
+import org.sonarsource.sonarlint.core.client.api.standalone.StandaloneSonarLintEngine;
+
+class RuleDescriptionHttpServer extends NanoHTTPD {
+
+  private final StandaloneSonarLintEngine engine;
+
+  RuleDescriptionHttpServer(int port, StandaloneSonarLintEngine engine) {
+    super(port);
+    this.engine = engine;
+  }
+
+  @Override
+  public Response serve(IHTTPSession session) {
+    String ruleKey = session.getParms().get("ruleKey");
+    if (ruleKey == null) {
+      return newFixedLengthResponse(Status.BAD_REQUEST, NanoHTTPD.MIME_PLAINTEXT, "Missing 'ruleKey' parameter");
+    } else {
+      RuleDetails ruleDetails;
+      try {
+        ruleDetails = engine.getRuleDetails(ruleKey);
+      } catch (IllegalArgumentException e) {
+        return newFixedLengthResponse(Status.NOT_FOUND, NanoHTTPD.MIME_PLAINTEXT, "No rules with key '" + ruleKey + "'");
+      }
+      try {
+        String ruleName = ruleDetails.getName();
+        String htmlDescription = ruleDetails.getHtmlDescription();
+        String type = ruleDetails.getType().toLowerCase(Locale.ENGLISH);
+        String typeImg64 = getAsBase64("/images/type/" + type + ".png");
+        String severity = ruleDetails.getSeverity().toLowerCase(Locale.ENGLISH);
+        String severityImg64 = getAsBase64("/images/severity/" + severity + ".png");
+        return newFixedLengthResponse("<!doctype html><html><head>" + css() + "</head><body><h1><big>"
+          + ruleName + "</big> (" + ruleKey
+          + ")</h1>"
+          + "<div>"
+          + "<img style=\"padding-bottom: 1px;vertical-align: middle\" width=\"16\" height=\"16\" alt=\"" + type + "\" src=\"data:image/gif;base64," + typeImg64 + "\">&nbsp;"
+          + clean(type)
+          + "&nbsp;"
+          + "<img style=\"padding-bottom: 1px;vertical-align: middle\" width=\"16\" height=\"16\" alt=\"" + severity + "\" src=\"data:image/gif;base64," + severityImg64
+          + "\">&nbsp;"
+          + clean(severity)
+          + "</div>"
+          + "<div class=\"rule-desc\">" + htmlDescription
+          + "</div></body></html>");
+      } catch (Exception e) {
+        StringWriter sw = new StringWriter();
+        PrintWriter pw = new PrintWriter(sw);
+        e.printStackTrace(pw);
+        return newFixedLengthResponse(Status.INTERNAL_ERROR, NanoHTTPD.MIME_PLAINTEXT, e.getMessage() + "\n" + sw.toString());
+      }
+    }
+  }
+
+  private static String getAsBase64(String image) throws IOException {
+    InputStream resourceAsStream = ServerMain.class.getResourceAsStream(image);
+    if (resourceAsStream == null) {
+      throw new IllegalStateException("Unable to load image " + image);
+    }
+    return Base64.getEncoder().encodeToString(IOUtils.toByteArray(resourceAsStream));
+  }
+
+  private static String clean(String txt) {
+    return StringUtils.capitalize(txt.toLowerCase(Locale.ENGLISH).replace("_", " "));
+  }
+
+  private static String css() {
```

Comment:
I think it would be better to add this as a resource which would be loaded once statically